sdk/core/core/src/az_http_request_builder.c

```diff
@@ -3,48 +3,188 @@
 
 #include <az_http_request_builder.h>
 
+#include <az_contract.h>
+#include <az_str.h>
+
+#include <assert.h>
+
 #include <_az_cfg.h>
 
+az_const_span const AZ_HTTP_METHOD_VERB_GET = AZ_CONST_STR("GET");
+az_const_span const AZ_HTTP_METHOD_VERB_HEAD = AZ_CONST_STR("HEAD");
+az_const_span const AZ_HTTP_METHOD_VERB_POST = AZ_CONST_STR("POST");
+az_const_span const AZ_HTTP_METHOD_VERB_PUT = AZ_CONST_STR("PUT");
+az_const_span const AZ_HTTP_METHOD_VERB_DELETE = AZ_CONST_STR("DELETE");
+az_const_span const AZ_HTTP_METHOD_VERB_TRACE = AZ_CONST_STR("TRACE");
+az_const_span const AZ_HTTP_METHOD_VERB_OPTIONS = AZ_CONST_STR("OPTIONS");
+az_const_span const AZ_HTTP_METHOD_VERB_CONNECT = AZ_CONST_STR("CONNECT");
+az_const_span const AZ_HTTP_METHOD_VERB_PATCH = AZ_CONST_STR("PATCH");
+
+AZ_INLINE az_pair * get_headers_start(
+    az_span const buffer,
+    int16_t const max_url_size,
+    int16_t const max_headers) {
+  uint8_t * const buffer_begin = buffer.begin;
+  uint8_t * const buffer_end = buffer.begin + buffer.size;
+
+  uint8_t * const headers_start
+      = (uint8_t *)((uintptr_t)(buffer_begin + max_url_size + 7) & ~(uintptr_t)7);
+
+  uint8_t * const headers_end = (uint8_t *)((az_pair *)headers_start + max_headers);
+
+  return headers_end <= buffer_end && buffer_begin < buffer_end && headers_start < headers_end
+      ? (az_pair *)headers_start
+      : NULL;
+}
+
 az_result az_http_request_builder_init(
     az_http_request_builder * const p_hrb,
     az_span const buffer,
-    int16_t const max_url_size,
     az_const_span const method_verb,
-    az_const_span const initial_url) {
-  (void)p_hrb;
-  (void)buffer;
-  (void)max_url_size;
-  (void)method_verb;
-  (void)initial_url;
+    az_const_span const initial_url,
+    uint16_t const max_url_size,
+    uint16_t const max_headers) {
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_span_is_valid(buffer) || !az_const_span_is_valid(method_verb)
+      || !az_const_span_is_valid(initial_url) || max_url_size < initial_url.size) {
+    return AZ_ERROR_ARG;
+  }
+
+  if (buffer.size < max_url_size
+      || (max_headers > 0 && get_headers_start(buffer, max_url_size, max_headers) == NULL)) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+
+  AZ_RETURN_IF_FAILED(az_span_set(buffer, '\0'));
```

Comment:
is this a requirement?  why do we want to clean the buffer like this ?

Reply:
Not a requirement, we don't have to do this.